File: daily/Binary Tree/103-Binary-tree-zigzag-level-order-traversal.go
package daily

func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	results := [][]int{}
	queue := []*TreeNode{root}
	level := 0

	for len(queue) > 0 {
		levelLength := len(queue)
		levels := []int{}

		for i := 0; i < levelLength; i++ {
			node := queue[0]
			queue = queue[1:] // popleft

			levels = append(levels, node.Val)

			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		if level%2 != 0 {
			left, right := 0, len(levels)-1
			for left < right {
				levels[left], levels[right] = levels[right], levels[left]
				left++
				right--
			}
		}

		results = append(results, levels)
		level++
	}
	return results
}

File: daily/Binary Tree/103-Binary-tree-zigzag-level-order-traversal.py
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        results = []
        queue = deque([root])
        level = 0
        while queue:
            length_level = len(queue)
            levels = []
            for _ in range(length_level):
                node = queue.popleft()
                levels.append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            if level % 2 != 0:
                left, right = 0, len(levels) - 1
                while left < right:
                    levels[left], levels[right] = levels[right], levels[left]
                    left += 1
                    right -= 1
            results.append(levels)
            level += 1
        return results
        